url46a88cc932_toby1984_tinyc_tgz-pJ8-de_codesourcery_dcpu16_compiler_CompilerJ8/src/de/codesourcery/dcpu16/codegeneration/CPUModel.java

```diff
@@ -0,0 +1,449 @@
+        public SlotTarget indirect()
+        {
+            if ( indirect ) {
+                return this;
+            }
+            return new RegisterTarget(register,true);
+        }
+
+        public boolean isIndirect()
+        {
+            return indirect;
+        }
+    }
+
+    protected final class MemoryTarget extends SlotTarget 
+    {
+        private final String label;
+        private final boolean indirect;
+
+        public MemoryTarget(String label,boolean indirect)
+        {
+            this.label = label;
+            this.indirect = indirect;
+        }
+        
+        @Override
+        public boolean isIndirect() {
+            return indirect;
+        }
+
+        @Override
+        public String toOperandString()
+        {
+            if ( indirect ) {
+                return "["+label+"]";
+            }
+            return label;
+        }        
+
+        @Override
+        public SlotTarget indirect()
+        {
+            if ( indirect ) {
+                return this;
+            }
+            return new MemoryTarget(label,true);
+        }
+    }    
+
+    protected class FunctionParameterTarget extends SlotTarget 
+    {
+        private final int index; // index relative to FRAME_POINTER
+
+        public FunctionParameterTarget(int index)
+        {
+            this.index=index;
+        }
+        
+        @Override
+        public boolean isIndirect() {
+            return true;
+        }        
+
+        @Override
+        public String toOperandString()
+        {
+            final String reg = FRAME_POINTER.toOperandString();
+            if ( index > 0 ) 
+            {
+                return "["+reg+"+"+index+"]";                
+            } 
+            if ( index < 0 ) {
+                return "["+reg+"-"+(-index)+"]";
+            }
+            return "["+reg+"]";
+        }        
+
+        @Override
+        public SlotTarget indirect()
+        {
+            return this;
+        }
+    }
+
+    protected final class ConstantValueTarget extends SlotTarget 
+    {
+        private final long value;
+
+        public ConstantValueTarget(long value)
+        {
+            this.value=value;
+        }
+
+        @Override
+        public boolean isIndirect() {
+            throw new UnsupportedOperationException("Literal value");
+        }
+        
+        @Override
+        public String toOperandString()
+        {
+            return Long.toString(value);
+        }        
+
+        @Override
+        public SlotTarget indirect()
+        {
+            throw new UnsupportedOperationException("Indirection not possible for constant value "+this);
+        }
+    }     
+    
+    private AssemblyCommand queue;
+    
+    protected final void flushCommandQueue() 
+    {
+        if ( queue != null ) 
+        {
+            output( queue );
+            queue = null;
+        }
+    }
+    
+    protected final void output(AssemblyCommand currentCmd) 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 364-373, lines 387-402 into separate methods.